### PR TITLE
Better support for touch events

### PR DESCRIPTION
**Pull request by ayuscode**
Better support for touch events (touchstart and mousedown) added.

For check if canvas support the event listener I added Emit code:
`[<Emit("$0 in $1")>]`
`let checkIn (listener: string) (o: obj) : bool = jsNative`
`if (checkIn "ontouchstart" canvas) then`

to simulate JS code `(if "ontouchstart" in canvas...)`

I'am wondering if there is a better option using Fable and F# code.

### CODE REVIEW

File: src/funsnake/App.fs [hunk 1]
```diff
@@ -57,6 +57,25 @@ let skipLast xs =
...
 
+// Touch / Mouse listeners
+[<Emit("$0 in $1")>]
+let checkIn (listener: string) (o: obj) : bool = jsNative
+
+let setTouchListener = 
+  if (checkIn "ontouchstart" canvas) then
+      // Capture touchstart for mobile / touch devices
+      canvas.addEventListener("touchstart", unbox(fun e ->
+          let touchEv= unbox<TouchEvent> e
```
Comment (alfonsogarciacaro) [inline src/funsnake/App.fs:68]: We can remove `unbox` by using the following code:

```fsharp
      canvas.addEventListener("touchstart", !^Browser.EventListener(fun e ->
          let touchEv = e :?> TouchEvent
```

File: src/funsnake/App.fs [hunk 2]
```diff
@@ -57,6 +57,25 @@ let skipLast xs =
...
+          let ev = touchEv.changedTouches.Item 0
+          touch <- (ev.pageX - canvas.offsetLeft, ev.pageY - canvas.offsetTop)))
+  else
+      // Capture mousedown for browsers
+      canvas.addEventListener("mousedown", unbox(fun e -> 
+          let ev = unbox<MouseEvent> e
```
Comment (alfonsogarciacaro) [inline src/funsnake/App.fs:74]: Same :)

```fsharp
      canvas.addEventListener("mousedown", !^Browser.EventListener(fun e ->
          let ev = e :?> MouseEvent
```

File: src/funsnake/App.fs [hunk 3]
```diff
@@ -182,12 +202,10 @@ let main() =
...
         :> obj)
 
-    // Capture MSPointerDown event for IE to move the snake (touch events for Chrome & iOS are in the html code)
-    canvas.addEventListener("pointerdown", unbox(fun e ->
-        let ev= unbox<MSPointerEvent> e
-        touch <- (ev.pageX - canvas.offsetLeft, ev.pageY - canvas.offsetTop)))
-
-    // Draw the walls only once
+    // Capture touch events
+    setTouchListener
```
Comment (alfonsogarciacaro) [inline src/funsnake/App.fs:206]: `setTouchListener` is a value, so it's eagerly evaluated and this doesn't do anything. You could convert it into a function call `setTouchListener()` (also changing the declaration to `let setTouchListener () =`)

**Review by alfonsogarciacaro — CHANGES_REQUESTED** [review]

**Comment by alfonsogarciacaro** [comment]
Hi Alberto! Long time no see :) It's great that you're updating your sample, thank you! 😻 

> I'am wondering if there is a better option using Fable and F# code.

Good question. Hmm, we don't have anything special in Fable to check for `undefined`. Options are being erased in Fable among other reasons to emulate optional members in Typescript, and because of this null checks in Fable use non-strict equality`==` which doesn't make a difference between `null` and `undefined` so it won't be useful here. I'm not sure if there's a more _elegant_ solution, but yours will work for now 👍 

I left some comments, could you please check? Also it'd be great if you could remove package-lock.json from the PR (which adds more than 5000 lines :wink:) as we're using yarn in this repo.

**Comment by ayuscode** [comment]
Hi Alfonso!!! Yes....Very long time... 
I'm glad to see Fable success and your incredible work and contribution to F# community!
I hope to have more time to contribute to Fable from now on.

I'm going to review your requested changes and create another pull request.

Thanks for your comments

**Pull request closed without merging**